### PR TITLE
Lax'd/mapped silhouette and squareform pdist

### PR DESCRIPTION
**Pull request by adamgayoso**
We can ironically reduce complexity by making each cell type the same shape (cells by features). To do so we pad zeros so each cell type has the number of cells of the most frequent cell type. Then we mask sums and means to ignore the padded zeros when necessary.

Compile time is way down now, on the gaming computer it's about 1/5 of the previous time.

### CODE REVIEW

**Comment by adamgayoso** [comment]
@martinkim0 you can use the pdist here in the k-means stuff

File: src/scib_metrics/utils/_silhouette.py [hunk 1]
```diff
@@ -108,6 +139,32 @@ def silhouette_samples(X: np.ndarray, labels: np.ndarray) -> np.ndarray:
...
+    # Compute intra-cluster distances
+    intra_dist_shuffled = _intra_cluster_distances(X).ravel()[cumulative_mask]
+    # Now unshuffle it
+    intra_dist = jnp.zeros_like(intra_dist_shuffled)
+    intra_dist = intra_dist.at[remapped_inds].set(intra_dist_shuffled)
+
+    # Compute nearest-cluster distances
+    inter_dist_per_label = jnp.inf * jnp.ones((n_labels, X.shape[1]))
+    inter_inds = jnp.triu_indices(n_labels, k=1)
+    inter_dist_a_b, inter_dist_b_a = _nearest_cluster_distances(X, inter_inds)
```
Comment (justjhong) [inline src/scib_metrics/utils/_silhouette.py:154]: you can unpack a single jax array like this?

**Review by justjhong — COMMENTED** [review]
some comments, LG otherwise

File: src/scib_metrics/utils/_silhouette.py [hunk 2]
```diff
@@ -1,90 +1,119 @@
...
-    intra_dist = intra_dist.at[labels_inds].set(values)
-    return intra_dist
+@jax.jit
+def _intra_cluster_distances_block(subset: jnp.ndarray) -> jnp.ndarray:
+    mask = subset.sum(1) != 0
+    full_mask = jnp.outer(mask, mask)
+    distances = pdist_squareform(subset)
+    per_cell_sum = jnp.where(full_mask, distances, 0).sum(axis=1)
+    real_cells_in_subset = mask.sum()
+    per_cell_mean = per_cell_sum / (real_cells_in_subset - 1)
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply src/scib_metrics/utils/_silhouette.py:42]: distance to self is 0, and want the mean distance over all other cells in the cluster

File: src/scib_metrics/utils/_dist.py [hunk 3]
```diff
@@ -28,3 +28,30 @@ def cdist(x: np.ndarray, y: np.ndarray) -> jnp.ndarray:
...
+        Array of shape (n_cells, n_cells)
+    """
+    # TODO(adamgayoso): Figure out how to speed up something like this
+    # n_cells = X.shape[0]
+    # inds = jnp.triu_indices(n_cells)
+    # dist_mat = jnp.zeros((n_cells, n_cells))
+    # dist_mat = dist_mat.at[inds].set(
+    #     jax.vmap(lambda i, j, X: _euclidean_distance(X[i], X[j]), in_axes=(0, 0, None))(*inds, jnp.asarray(X))
+    # )
+    # dist_mat = jnp.maximum(dist_mat, dist_mat.T)
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply src/scib_metrics/utils/_dist.py:56]: yeah something is way off, I think it tries to expand the X matrix a bunch and runs out of memory.

It's also really slow to compile

File: src/scib_metrics/utils/_silhouette.py [hunk 4]
```diff
@@ -1,90 +1,119 @@
...
+def _nearest_cluster_distance_block(subset_a: jnp.ndarray, subset_b: jnp.ndarray) -> Union[jnp.ndarray, jnp.ndarray]:
+    mask_a = subset_a.sum(1) != 0
+    mask_b = subset_b.sum(1) != 0
+    full_mask = jnp.outer(mask_a, mask_b)
     distances = cdist(subset_a, subset_b)
-    values_a = distances.mean(axis=1)
-    values_b = distances.mean(axis=0)
+    masked_distances = jnp.where(full_mask, distances, 0)
+    values_a = masked_distances.sum(axis=1)
+    values_b = jnp.where(full_mask, distances, 0).sum(axis=0)
```
Comment (adamgayoso) [inline src/scib_metrics/utils/_silhouette.py:67]:
```suggestion
    values_b = masked_distances.sum(axis=0)
```

[the remark this answers is not in the document]
Comment (adamgayoso) [reply src/scib_metrics/utils/_silhouette.py:67]: yes

File: src/scib_metrics/utils/_silhouette.py (hunk 1 above)
Comment (adamgayoso) [reply src/scib_metrics/utils/_silhouette.py:154]: the block function is returning a tuple

Comment (justjhong) [reply src/scib_metrics/utils/_silhouette.py:154]: oh, then the return typing is wrong

File: src/scib_metrics/utils/_silhouette.py [hunk 5]
```diff
@@ -1,90 +1,119 @@
...
+    per_cell_sum = jnp.where(full_mask, distances, 0).sum(axis=1)
+    real_cells_in_subset = mask.sum()
+    per_cell_mean = per_cell_sum / (real_cells_in_subset - 1)
+    return per_cell_mean
+
+
+# @jax.jit
+def _nearest_cluster_distances(X: jnp.ndarray, inds: jnp.ndarray = None) -> jnp.ndarray:
+    """Calculate the mean nearest-cluster distance for observation i."""
+    carry = 0
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply src/scib_metrics/utils/_silhouette.py:49]: should use map

File: src/scib_metrics/utils/_silhouette.py (hunk 1 above)
Comment (adamgayoso) [reply src/scib_metrics/utils/_silhouette.py:153]: fixed

File: src/scib_metrics/utils/_silhouette.py [hunk 6]
```diff
@@ -1,90 +1,118 @@
...
+    full_mask = jnp.outer(mask, mask)
+    distances = pdist_squareform(subset)
+    per_cell_sum = jnp.where(full_mask, distances, 0).sum(axis=1)
+    real_cells_in_subset = mask.sum()
+    per_cell_mean = per_cell_sum / (real_cells_in_subset - 1)
+    return per_cell_mean
+
+
+# @jax.jit
+def _nearest_cluster_distances(X: jnp.ndarray, inds: jnp.ndarray = None) -> jnp.ndarray:
```
Comment (justjhong) [inline src/scib_metrics/utils/_silhouette.py:47]:
```suggestion
def _nearest_cluster_distances(X: jnp.ndarray, inds: Optional[Tuple[jnp.ndarray, jnp.ndarray]] = None) -> Tuple[jnp.ndarray, jnp.ndarray]:
```

**Review by justjhong — APPROVED** [review]

File: src/scib_metrics/utils/_silhouette.py (hunk 6 above)
Comment (adamgayoso) [inline src/scib_metrics/utils/_silhouette.py:47]:
```suggestion
def _nearest_cluster_distances(X: jnp.ndarray, inds: jnp.ndarray) -> Tuple[jnp.ndarray, jnp.ndarray]:
```

**Pull request merged**